rust/chains/tw_polkadot/src/extrinsic.rs

```diff
@@ -0,0 +1,554 @@
+}
+
+impl ExtrinsicEncoder {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    fn from_encoded(encoded: Vec<u8>) -> Self {
+        Self { encoded }
+    }
+
+    pub fn add(self, t: impl ToScale) -> Self {
+        let mut encoded = self.encoded;
+        t.to_scale_into(&mut encoded);
+        Self::from_encoded(encoded)
+    }
+
+    pub fn add_mut(&mut self, t: impl ToScale) -> &mut Self {
+        t.to_scale_into(&mut self.encoded);
+        self
+    }
+
+    pub fn finalize(self) -> Vec<u8> {
+        self.encoded
+    }
+}
+
+// `Extrinsic` is (for now) just a lightweight wrapper over the actual protobuf object.
+// In the future, we will refine the latter to let the caller specify arbitrary extrinsics.
+
+#[derive(Debug, Clone)]
+pub struct Extrinsic<'a> {
+    inner: Proto::SigningInput<'a>,
+}
+
+impl<'a> Extrinsic<'a> {
+    pub fn from_input(input: Proto::SigningInput<'a>) -> Self {
+        Self { inner: input }
+    }
+
+    fn get_call_index_for_network(network: NetworkId, key: &str) -> EncodeResult<(u8, u8)> {
+        CALL_INDICES_BY_NETWORK
+            .get(&network)
+            .and_then(|table| table.get(key))
+            .cloned()
+            .ok_or(EncodeError::MissingCallIndicesTable)
+    }
+
+    fn get_custom_call_index(civ: &Option<Proto::CallIndices>) -> EncodeResult<(u8, u8)> {
+        if let Some(CallIndicesVariant::custom(c)) = civ.as_ref().map(|i| &i.variant) {
+            if c.module_index > 0xff || c.method_index > 0xff {
+                return Err(EncodeError::InvalidCallIndex);
+            }
+            return Ok((c.module_index as u8, c.method_index as u8));
+        }
+        Err(EncodeError::MissingCallIndicesTable)
+    }
+
+    fn get_custom_call_index_or_network(
+        &self,
+        key: &str,
+        civ: &Option<Proto::CallIndices>,
+    ) -> EncodeResult<(u8, u8)> {
+        Self::get_custom_call_index(civ).or_else(|_| {
+            let network = NetworkId::try_from(self.inner.network as u16)
+                .map_err(|_| EncodeError::InvalidNetworkId)?;
+
+            Self::get_call_index_for_network(network, key)
+        })
+    }
+
+    fn should_encode_raw_account(&self) -> bool {
+        if self.inner.multi_address {
+            return false;
+        }
+
+        let (network, spec) = (
+            NetworkId::try_from(self.inner.network as u16),
+            self.inner.spec_version,
+        );
+
+        match (network, spec) {
+            (Ok(NetworkId::POLKADOT), _) if spec >= POLKADOT_MULTI_ADDRESS_SPEC => false,
+            (Ok(NetworkId::KUSAMA), _) if spec > KUSAMA_MULTI_ADDRESS_SPEC => false,
+            _ => true,
+        }
+    }
+
+    fn encode_transfer(&self, t: &Transfer) -> EncodeResult<Vec<u8>> {
+        let call_index =
+            self.get_custom_call_index_or_network(BALANCE_TRANSFER, &t.call_indices)?;
+        let address =
+            SS58Address::from_str(&t.to_address).map_err(|_| EncodeError::InvalidAddress)?;
```

Comment:
Recently, we added error contexts to add some useful info to the `SigningOutput.error_message` string: https://github.com/trustwallet/wallet-core/pull/3806
To be able to add an error context, please consider changing the EncodeResult type to `Result<T, TWError<EncodeError>>`.